src/ui/display.c

```diff
@@ -74,99 +69,186 @@ UX_STEP_CB(ux_display_reject_step,
+        total_outputs--;
+        if (G_context.tx_info.decoded_output.index > G_context.tx_info.change_output_index) {
+            fake_output_index--;
+        }
+    }
+    itoa(fake_output_index, g_output_index, 10);
+    uint8_t len = strlen(g_output_index);
+    g_output_index[len++] = '/';
+    itoa(total_outputs, g_output_index+len, 10);
+
+    // set g_address
+    memset(g_address, 0, sizeof(g_address));
+    char b58address[B58_ADDRESS_LEN] = {0};
+    uint8_t address[ADDRESS_LEN] = {0};
+    address_from_pubkey_hash(G_context.tx_info.decoded_output.pubkey_hash, address);
+    base58_encode(address, ADDRESS_LEN, b58address, B58_ADDRESS_LEN);
+    snprintf(g_address, sizeof(g_address), "%.*H", sizeof(b58address), b58address);
+
+    // set g_ammount (HTR value)
+    memset(g_amount, 0, sizeof(g_amount));
+    strcpy(g_amount, "HTR ");
+    format_value(G_context.tx_info.decoded_output.value, g_amount+4);
+
+    g_validate_callback = cb; // decode and show next until need more
+    ux_flow_init(0, ux_display_tx_output_flow, NULL);
+
+    return 0;
+}
+
+/* FLOW to display confirm sign tx:
+ *  #1 screen : eye icon + "Send Transaction?"
+ *  #2 screen : approve button
+ *  #3 screen : reject button
+ */
+UX_FLOW(ux_display_confirm_tx_flow,
+        &ux_display_confirm_sign_step,  // -> <eye> Send Transaction?
+        &ux_display_approve_step,       // -> return SW_OK
+        &ux_display_reject_step);       // -> return reject
+
+int ui_display_tx_confirm() {
+    if (G_context.req_type != CONFIRM_TRANSACTION || G_context.state != STATE_PARSED) {
+        G_context.state = STATE_NONE;
+        return io_send_sw(SW_BAD_STATE);
+    }
+
+    g_validate_callback = &ui_action_tx_confirm; // set state_approved and send sw_ok
+
+    ux_flow_init(0, ux_display_confirm_tx_flow, NULL);
+
+    return 0;
+}
+
+// Get XPUB: ui_display_xpub_confirm
+
+/* FLOW to display confirm access to XPUB:
+ *  #1 screen: eye icon + "Confirm Access?"
+ *  #2 screen: display BIP32 Path
+ *  #3 screen: approve button
+ *  #4 screen: reject button
+ */
+UX_FLOW(ux_display_xpub_flow,
+        &ux_display_confirm_step,  // <eye> Confirm Access?
+        &ux_display_path_step,          // bip32 path
+        &ux_display_approve_step,       // accept => send xpub in response
+        &ux_display_reject_step);       // reject => return error
+
+int ui_display_xpub_confirm() {
+    if (G_context.req_type != CONFIRM_XPUB || G_context.state != STATE_NONE) {
         G_context.state = STATE_NONE;
         return io_send_sw(SW_BAD_STATE);
     }
+    uint32_t full_path[MAX_BIP32_PATH];
+    init_bip32_full_path(G_context.bip32_path.path, G_context.bip32_path.length, full_path);
 
     memset(g_bip32_path, 0, sizeof(g_bip32_path));
-    if (!bip32_path_format(G_context.bip32_path,
-                           G_context.bip32_path_len,
+
+    if (!bip32_path_format(full_path,
+                           2+G_context.bip32_path.length,
```

Comment:
Why do we have this magic `2` here? As far as I understood, the bip32_path has only part of the length (and we have two other parts that are hardcoded). Is that correct? Maybe we should put it into a constant as close as possible to the hardcoded parts.